mixins/interactive-mixin.js

```diff
@@ -0,0 +1,115 @@
+import { LocalizeCoreElement } from '../helpers/localize-core-element.js';
+
+const keyCodes = {
+	ENTER: 13,
+	ESCAPE: 27
+};
+
+export const InteractiveMixin = superclass => class extends LocalizeCoreElement(superclass) {
+
+	static get properties() {
+		return {
+			_hasInteractiveAncestor: { state: true },
+			_interactive: { state: true }
+		};
+	}
+
+	static get styles() {
+		return css`
+			.interactive-container.focus-visible,
+			.interactive-container:focus-visible {
+				border-radius: 6px;
+				outline: 2px solid var(--d2l-color-celestine);
+				outline-offset: 2px;
+			}
+		`;
+	}
+
+	constructor() {
+		super();
+		this._hasInteractiveAncestor = false;
+		this._interactive = false;
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+
+		const parentGrid = findComposedAncestor(this.parentNode, node => {
+			return ((node.tagName === 'D2L-LIST' && node.grid) || (node.nodeType === Node.ELEMENT_NODE && node.getAttribute('role') === 'grid'));
+		});
+		this._hasInteractiveAncestor = (parentGrid !== null);
+	}
+
+	focus() {
+		if (!this.shadowRoot) return;
+		if (this._hasInteractiveAncestor && !this._interactive) forceFocusVisible(this.shadowRoot.querySelector('.interactive-container'));
+		else this._focusInteractiveContent();
+	}
+
+	_handleInteractiveContentFocusIn() {
+		this._interactive = true;
+	}
+
+	_handleInteractiveContentFocusOut(e) {
+		if (isComposedAncestor(this.shadowRoot.querySelector('.interactive-container-content'), e.relatedTarget)) return;
+		// focus moved out of the interactive content
+		this._interactive = false;
+	}
+
+	async _handleInteractiveEndFocus() {
+		// focus moved to trap-end either forwards from contents or backwards from outside - focus interactive toggle
+		this._interactive = false;
+		await this.updateComplete;
+		this.shadowRoot.querySelector('.interactive-container').focus();
+	}
+
+	async _handleInteractiveKeyDown(e) {
+		if (!this._hasInteractiveAncestor) return;
+		if (!this._interactive && e.keyCode === keyCodes.ENTER) {
+			this._interactive = true;
+			await this.updateComplete;
+			this.focus();
+		} else if (this._interactive && e.keyCode === keyCodes.ESCAPE) {
+			this._interactive = false;
+			await this.updateComplete;
+			this.shadowRoot.querySelector('.interactive-container').focus();
+		}
+	}
+
+	async _handleInteractiveStartFocus(e) {
+		if (e.relatedTarget === this.shadowRoot.querySelector('.interactive-container')) {
+			// focus moved to trap-start while non-interactive - focus next focusable after this component
+			const nextFocusable = getNextFocusable(this.shadowRoot.querySelector('.interactive-trap-end'));
+			if (nextFocusable) nextFocusable.focus();
+		} else {
+			// focus moved to trap-start backwards from within contents - toggle to non-interactive and apply focus
+			this._interactive = false;
+			await this.updateComplete;
+			this.shadowRoot.querySelector('.interactive-container').focus();
+		}
+	}
+
+	_renderInteractiveContainer(inner, label) {
```

Comment:
Should this method be "private" since it's the main entrypoint to the mixin? I know you're going of treating it more like a protected method... but can/should mixins have public methods if they're not going to collide with each other?

Reply:
Yeah, it probably should be public.  I don't think we've been entirely consistent with this.  In some cases I think they been marked at private because we don't want them to be used publicly, but rather only within our stuff.  In this case though, I don't think that's the case. An argument could be made they should still be public though.